dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java

```diff
@@ -1834,14 +1850,264 @@ private void updateTask(WorkflowProcessor processor) throws DotDataException{
+
+		try {
+			final List<Contentlet> contentlets = findContentletsToProcess(cleanedUpQuery,
+					workflowAssociatedStepsIds, user);
+			final Long skipsCount = computeSkippedContentletsCount(cleanedUpQuery, workflowAssociatedStepsIds,
+					user);
+			return distributeWorkAndProcess(action, user, contentlets, skipsCount);
+		} catch (Exception e) {
+			Logger.error(getClass(), "Error firing actions in bulk.", e);
+			throw new DotDataException(e);
+		}
+
+	}
+
+	/**
+	 * This method takes the contentlets that are the result of a selection made on the UI or the
+	 * product of a lucene Query The list then gets partitioned and distributed between the list of
+	 * available workers (threads).
+	 * @param action
+	 * @param user
+	 * @param contentlets
+	 * @param skipsCount
+	 * @return
+	 */
+	private BulkActionsResultView distributeWorkAndProcess(final WorkflowAction action,
+			final User user, final List<Contentlet> contentlets, final Long skipsCount) {
+		final DotSubmitter submitter = this.concurrentFactory
+				.getSubmitter(DotConcurrentFactory.DOT_SYSTEM_THREAD_POOL);
+
+		final AtomicLong successCount = new AtomicLong();
+		final AtomicLong failsCount = new AtomicLong();
+
+		final List<Future> futures = new ArrayList<>();
+
+		final List<List<Contentlet>> partitions = partitionContentletInput(contentlets);
+
+		for (final List<Contentlet> partition : partitions) {
+			futures.add(
+					submitter.submit(() -> fireBulkActionTasks(action, user, partition,
+							successCount::addAndGet, failsCount::addAndGet))
+			);
+		}
+
+		for (final Future future : futures) {
+			try {
+				future.get();
+			} catch (InterruptedException | ExecutionException e) {
+				Logger.error(this, e.getMessage(), e);
+			}
+		}
 
+		return new BulkActionsResultView(
+				successCount.get(),
+				skipsCount,
+				failsCount.get()
+		);
+	}
+
+
+	/**
+	 * Takes a input list of contentlets and creates sub groups to distribute the workload
+	 */
+	private List<List<Contentlet>> partitionContentletInput(final List<Contentlet> contentlets) {
+		final int maxThreads = Config.getIntProperty(MAX_THREADS_ALLOWED_TO_HANDLE_BULK_ACTIONS, MAX_THREADS_ALLOWED_TO_HANDLE_BULK_ACTIONS_DEFAULT);
+		final int partitionSize = Math
+				.max((contentlets.size() / maxThreads), 10);
+
+		Logger.info(getClass(),
+				String.format(
+						"Number of threads is limited to %d. Number of Contentlets to process is %d. Load will be distributed in bunches of %d ",
+						maxThreads, contentlets.size(),
+						partitionSize)
+		);
+		return Lists.partition(contentlets, partitionSize);
+
+	}
+
+	private void fireBulkActionTasks(final WorkflowAction action,
+			final User user,
+			final List<Contentlet> contentlets,
+			final Consumer<Long> successConsumer,
+			final Consumer<Long> failConsumer) {
+
+		for (final Contentlet contentlet : contentlets) {
+			try {
+				fireBulkActionTask(action, user, contentlet, successConsumer, failConsumer);
+			} catch (Exception e) {

```

Comment:
Yeah I agree.. it looks like an additional catch. That isn't doing much there. But I think This one is there to handle any exceptions that might arise from processing the @WrapInTransaction 
In case something rolls back the rest of the contentlets will still continue to get processed.

I'd rather keeping it and I will add some additional comments to explain why it is there.